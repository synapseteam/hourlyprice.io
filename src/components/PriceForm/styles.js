import { css } from "@emotion/react";
import { getStyle } from "utils/generic";

export const styles = {
  form: css`
    margin-top: 1rem;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    width: 100%;
  `,
  getStyle,
};
