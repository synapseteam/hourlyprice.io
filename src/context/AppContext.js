import { createContext, useReducer, useContext } from "react";

const AppThemeContext = createContext();

const initState = {
  darkMode: true,
};

const TOGGLE_THEME = "TOGGLE_THEME";

function reducer(state, action) {
  switch (action.type) {
    case TOGGLE_THEME:
      return {
        ...state,
        darkMode: !state.darkMode,
      };
    default:
      throw Error("impossible case while switching Theme");
  }
}

export const AppThemeProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initState);
  const value = [state, dispatch];

  return (
    <AppThemeContext.Provider value={value}>
      {children}
    </AppThemeContext.Provider>
  );
};

export const useAppThemeContext = () => {
  const context = useContext(AppThemeContext);

  if (!context) {
    throw Error("useAppThemeContext should be used within AppThemeProvider");
  }

  return context;
};

export const toggleTheme = () => {
  return { type: TOGGLE_THEME, payload: null };
};

export default AppThemeContext;
