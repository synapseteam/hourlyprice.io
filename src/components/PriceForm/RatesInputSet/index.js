/** @jsxImportSource @emotion/react */
import uniqid from "uniqid";
import PropTypes from "prop-types";

import { handlePriceChange } from "utils/generic";

import { styles } from "./styles";

export default function RatesInputSet({
  register,
  chosenCurrency,
  allCurrencies,
  errors,
}) {
  const allCurrenciesNames = allCurrencies.map((el) => ({ name: el.name }));

  return (
    <div css={styles.ratesInputsContainer}>
      {allCurrenciesNames.map(({ name }) => {
        return (
          <input
            css={styles.ratesInput}
            {...register(name)}
            key={uniqid()}
            inputMode="decimal"
            type="text"
            autoComplete="off"
            placeholder={name}
            onChange={handlePriceChange}
          />
        );
      })}
    </div>
  );
}

RatesInputSet.propTypes = {
  register: PropTypes.func.isRequired,
  chosenCurrency: PropTypes.string.isRequired,
  allCurrencies: PropTypes.arrayOf(PropTypes.object),
  erorrs: PropTypes.object,
};
