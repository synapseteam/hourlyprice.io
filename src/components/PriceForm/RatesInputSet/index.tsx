/** @jsxImportSource @emotion/react */
import { FC } from "react";
import { ICurrency } from "typescript/interfaces";
import uniqid from "uniqid";
import { FieldValues, UseFormRegister } from "react-hook-form";

import { handlePriceChange } from "utils/generic";

import { styles } from "./styles";

interface IRatesInputSet {
  register: UseFormRegister<FieldValues>;
  chosenCurrency: string;
  allCurrencies: ICurrency[];
  error: Record<string, string>;
}

const RatesInputSet: FC<IRatesInputSet> = ({ register, allCurrencies }) => {
  const allCurrenciesNames = allCurrencies.map((el: { name: string }) => ({
    name: el.name,
  }));

  return (
    <div css={styles.ratesInputsContainer}>
      {allCurrenciesNames.map(({ name }) => {
        return (
          <input
            css={styles.ratesInput}
            {...register(name)}
            key={uniqid()}
            inputMode="decimal"
            type="text"
            autoComplete="off"
            placeholder={name}
            onChange={handlePriceChange}
          />
        );
      })}
    </div>
  );
};

export default RatesInputSet;
