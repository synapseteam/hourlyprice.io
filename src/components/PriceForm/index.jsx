/** @jsxImportSource @emotion/react */
import { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import PropTypes from "prop-types";

import Input from "components/UI/Input";
import Select from "components/UI/Select";
import RatesInputSet from "components/PriceForm/RatesInputSet";
import Button from "components/UI/Button";
import { submitFieldsData, setRequestErr } from "store/actions/generic";
import { getNewRatesThunkCreator, setManualRates } from "store/actions/rates";
import { formSchema, ratesSources } from "configure";
import {
  convertStrTimeToNum,
  handleTimeChange,
  handlePriceChange,
} from "utils/generic";
import { ratesUpdatingTimeFrame } from "configure";
import { useCustomTranslation } from "i18n";

import { styles } from "./styles";

export default function PriceForm({ children }) {
  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors },
  } = useForm({
    resolver: yupResolver(formSchema),
  });

  const [t] = useCustomTranslation();

  const allCurrencies = useSelector((state) => state.rates.allCurrencies);
  const ratesSource = useSelector((state) => state.rates.ratesSource);
  const timeStampCurrenciesUpdated = useSelector(
    (state) => state.rates.updatedAt
  );
  const allCurrenciesNames = allCurrencies.map((el) => el.name);

  const dispatch = useDispatch();

  const [chosenCurrency, setChosenCurrency] = useState("USD");
  const [chosenRatesSource, setChosenRatesSource] = useState("MasterCard");

  const [keepAPIRatesCache, setkeepAPIRatesCache] = useState(true);

  function handleListChange(e) {
    const { name, value } = e.target;
    if (name === "currency") {
      setChosenCurrency((prev) => value);
      setValue(value, 1);
    }
    if (name === "ratesSource") {
      setChosenRatesSource((prev) => value);
    }
    return;
  }

  function updateRatesIfCacheExpired() {
    const timeStampNow = new Date();
    const timePassAfterRatesUpdated =
      (timeStampNow - timeStampCurrenciesUpdated) / 1000;

    if (
      timePassAfterRatesUpdated >= ratesUpdatingTimeFrame ||
      !keepAPIRatesCache
    ) {
      dispatch(getNewRatesThunkCreator());
    }
  }

  useEffect(() => {
    if (ratesSource === "Manual") {
      setkeepAPIRatesCache((prev) => false);
    }
    if (ratesSource === "MasterCard") {
      setkeepAPIRatesCache((prev) => true);
    }
  }, [ratesSource]);

  const onSubmit = ({
    price,
    time: timeString,
    currency,
    ratesSource,
    ...manualRates
  }) => {
    dispatch(setRequestErr(false));

    const time = convertStrTimeToNum(timeString);

    if (ratesSource === "MasterCard") {
      updateRatesIfCacheExpired();
    }

    if (ratesSource === "Manual") {
      dispatch(setManualRates(manualRates));
    }

    dispatch(submitFieldsData({ price, time, currency }));
  };

  useEffect(() => {
    setValue("UAH", "");
    setValue("RUB", "");
    setValue("EUR", "");
    setValue("USD", "");
    setValue(chosenCurrency, 1);
  }, [chosenCurrency]);

  return (
    <form css={styles.form} onSubmit={handleSubmit(onSubmit)}>
      <div>
        <Input
          inputName="price"
          register={register}
          labelName={t("labelPrice")}
          placeholder="20.30"
          changeHandler={handlePriceChange}
          errors={errors}
        />
        <Input
          inputName="time"
          register={register}
          labelName={t("labelTime")}
          placeholder={t("timePlaceholder")}
          changeHandler={handleTimeChange}
          errors={errors}
        />
        <Select
          labelName={t("labelCurrency")}
          inputName="currency"
          register={register}
          changeHandler={handleListChange}
          value={chosenCurrency}
          optionsArr={allCurrenciesNames}
          errors={errors}
        />
        <Select
          labelName={t("labelExchangeRate")}
          inputName="ratesSource"
          register={register}
          changeHandler={handleListChange}
          value={chosenRatesSource}
          optionsArr={ratesSources}
          errors={errors}
        />
        {chosenRatesSource === "Manual" && (
          <RatesInputSet
            register={register}
            allCurrencies={allCurrencies}
            chosenCurrency={chosenCurrency}
            errors={errors}
          />
        )}
        <Button />
      </div>
    </form>
  );
}

PriceForm.propTypes = {
  id: PropTypes.string,
};

PriceForm.defaultProps = {
  id: "calc-form",
};
