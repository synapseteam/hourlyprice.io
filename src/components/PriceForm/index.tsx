/** @jsxImportSource @emotion/react */
import { useState, useEffect, FC, ChangeEvent } from "react";
import { useAppDispatch, useAppSelector } from "store/hooks";
import { useForm, SubmitHandler, FieldValues } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import * as yup from "yup";

import InputLabel from "components/UI/InputLabel";
import Select from "components/UI/Select";
import RatesInputSet from "components/PriceForm/RatesInputSet";
import Button from "components/UI/Button";
import { ratesSources } from "configure";
import {
  convertStrTimeToNum,
  handleTimeChange,
  handlePriceChange,
} from "utils/generic";
import { ratesUpdatingTimeFrame } from "configure";
import { useCustomTranslation } from "i18n";
import { transformRatesResponse } from "utils/generic";
import { modifyFields, setRequestErr } from "features/generic";
import { fetchRates } from "features/rates";
import { setManualRates as setManualRates2 } from "features/rates";
import {
  clearFields,
  setInvoiceItemAdded,
  setInvoiceFull,
} from "features/generic";
import { MILISEC_IN_ONE_SEC } from "utils/constants";

import { styles } from "./styles";

const PriceForm: FC = (): JSX.Element => {
  const [t] = useCustomTranslation();

  const formSchema = yup.object().shape({
    price: yup.number().typeError(t("priceError")).positive().required(),
    time: yup.string().required(t("timeError")),
    currency: yup.string().required().oneOf(["USD", "UAH", "EUR"]),
  });
  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors },
  } = useForm<FieldValues>({
    resolver: yupResolver(formSchema),
  });

  const allCurrencies = useAppSelector((state) => state.rates.allCurrencies);
  const ratesSource = useAppSelector((state) => state.rates.ratesSource);
  const timeStampCurrenciesUpdated = useAppSelector(
    (state) => state.rates.updatedAt
  );
  const { price, time } = useAppSelector((state) => state.generic.fields);

  const allCurrenciesNames = allCurrencies.map((el) => {
    return { name: el.name, value: el.name };
  });

  const dispatch = useAppDispatch();
  const [chosenCurrency, setChosenCurrency] = useState("USD");
  const [chosenRatesSource, setChosenRatesSource] = useState("masterCard");

  const [keepAPIRatesCache, setkeepAPIRatesCache] = useState(true);

  useEffect(() => {
    const currency = JSON.parse(localStorage.getItem("currency"));
    if (currency) {
      setChosenCurrency(currency);
    }
  }, []);

  function handleListChange(event: ChangeEvent<HTMLSelectElement>) {
    const { name, value } = event.target;
    if (name === "currency") {
      setChosenCurrency(() => value);
      localStorage.setItem("currency", JSON.stringify(value));
      // eslint-disable-next-line no-magic-numbers
      setValue(value, 1);
    }
    if (name === "ratesSource") {
      setChosenRatesSource(() => value);
    }
    return;
  }
  function updateRatesIfCacheExpired() {
    const timeStampNow = new Date();
    const timePassAfterRatesUpdated =
      (Number(timeStampNow) - Number(timeStampCurrenciesUpdated)) /
      MILISEC_IN_ONE_SEC;

    if (
      timePassAfterRatesUpdated >= ratesUpdatingTimeFrame ||
      !keepAPIRatesCache
    ) {
      dispatch(fetchRates());
    }
  }

  useEffect(() => {
    if (ratesSource === "manual") {
      setkeepAPIRatesCache(() => false);
    }
    if (ratesSource === "masterCard") {
      setkeepAPIRatesCache(() => true);
    }
  }, [ratesSource]);

  const onSubmit: SubmitHandler<FieldValues> = ({
    price,
    time: timeString,
    currency,
    ratesSource,
    ...manualRates
  }) => {
    dispatch(setRequestErr(false));

    const time = convertStrTimeToNum(timeString);

    if (ratesSource === "masterCard") {
      updateRatesIfCacheExpired();
    }

    if (ratesSource === "manual") {
      const newRates = transformRatesResponse(manualRates);
      dispatch(setManualRates2(newRates));
    }

    dispatch(modifyFields({ price, time, currency }));
  };

  useEffect(() => {
    setValue("UAH", "");
    setValue("EUR", "");
    setValue("USD", "");
    // eslint-disable-next-line no-magic-numbers
    setValue(chosenCurrency, 1);
  }, [chosenCurrency]);

  const addItemToLocalStorage = () => {
    const invoiceItems = JSON.parse(localStorage.getItem("invoiceItems"));
    const invoiceObj = {
      description: "No description",
      price: price,
      time: time,
    };
    if (!invoiceItems) {
      const invoiceArray = [];
      invoiceArray.push(invoiceObj);
      localStorage.setItem("invoiceItems", JSON.stringify(invoiceArray));
      dispatch(clearFields());
    }
    if (invoiceItems && invoiceItems.length <= 9) {
      invoiceItems.push(invoiceObj);
      localStorage.setItem("invoiceItems", JSON.stringify(invoiceItems));
      dispatch(clearFields());
      dispatch(setInvoiceItemAdded(true));
    } else if (invoiceItems && invoiceItems.length >= 10) {
      dispatch(setInvoiceFull(true));
    }
  };

  const addToInvoiceButtonStyles = [
    styles.button,
    price ? styles.addToInvoiceButtonVisible : styles.addToInvoiceButton,
  ];

  const submitButtonStyles = price ? styles.button : styles.buttonWide;

  return (
    <form css={styles.form} onSubmit={handleSubmit(onSubmit)}>
      <div css={styles.currency}>
        <InputLabel
          inputName="price"
          register={register}
          labelName={t("labelPrice")}
          placeholder="0.0"
          changeHandler={handlePriceChange}
          errors={errors}
        />
        <Select
          labelName={t("labelCurrency")}
          inputName="currency"
          register={register}
          changeHandler={handleListChange}
          value={chosenCurrency}
          optionsArr={allCurrenciesNames}
          errors={errors}
        />
      </div>
      <InputLabel
        inputName="time"
        register={register}
        labelName={t("labelTime")}
        placeholder={t("timePlaceholder")}
        changeHandler={handleTimeChange}
        errors={errors}
      />
      <Select
        labelName={t("labelExchangeRate")}
        inputName="ratesSource"
        register={register}
        changeHandler={handleListChange}
        value={chosenRatesSource}
        optionsArr={ratesSources}
        errors={errors}
      />
      {chosenRatesSource === "manual" && (
        <RatesInputSet
          register={register}
          allCurrencies={allCurrencies}
          chosenCurrency={chosenCurrency}
          error={errors}
        />
      )}
      <div css={styles.buttons}>
        <div css={submitButtonStyles}>
          <Button type="submit">{t("btnResult")}</Button>
        </div>
        <div css={addToInvoiceButtonStyles}>
          <Button type="button" onClick={addItemToLocalStorage}>
            {t("addToInvoice")}
          </Button>
        </div>
      </div>
    </form>
  );
};

export default PriceForm;
