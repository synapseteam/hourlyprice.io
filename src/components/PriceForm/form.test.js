/**
 * @jest-environment jsdom
 */

import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";
import { AppThemeProvider } from "context/AppContext";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import thunkMiddleware from "redux-thunk";

import PriceForm from "components/PriceForm";

if (!global.React) {
  global.React = React;
}

const mockStore = configureStore([thunkMiddleware]);
const initState = {
  main: {
    fields: {
      price: "",
      time: "",
      currency: "USD",
    },
  },
  rates: {
    allCurrencies: [
      { name: "USD", rate: 1, symbol: "$", isoCode: "840" },
      { name: "EUR", rate: 1, symbol: "€", isoCode: "978" },
      { name: "UAH", rate: 1, symbol: "₴", isoCode: "980" },
      { name: "RUB", rate: 1, symbol: "₽", isoCode: "810" },
    ],
    updatedAt: "",
    ratesSource: "MasterCard",
  },
};

const context = {
  darkMode: true,
};
const value = [context];

const store = mockStore(initState);

const setup = () => {
  const { container } = render(
    <AppThemeProvider value={value}>
      <Provider store={store}>
        <PriceForm />
      </Provider>
    </AppThemeProvider>
  );

  const form = container.querySelector("form");

  const allLabels = form.querySelectorAll("label");

  const allLabelsTexts = [...allLabels].map((label) => label.textContent);

  let priceInput = screen.getByRole("textbox", {
    name: /price:/i,
  });

  const timeInput = screen.getByRole("textbox", {
    name: /time:/i,
  });

  const currencyInput = screen.getByRole("combobox", {
    name: /currency:/i,
  });

  const rateSource = screen.getByRole("combobox", {
    name: /exchange rate:/i,
  });

  return {
    form,
    allLabelsTexts,
    priceInput,
    timeInput,
    currencyInput,
    rateSource,
  };
};

test("Form component snapshot test", () => {
  const tree = renderer
    .create(
      <AppThemeProvider value={value}>
        <Provider store={store}>
          <PriceForm />
        </Provider>
      </AppThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Form component: user submit empty form", async () => {
  const { form, allLabelsTexts } = setup();

  expect(allLabelsTexts.length).toBe(4);

  await waitFor(() => {
    fireEvent.submit(form);
  });
  let allErrors = [...form.querySelectorAll(".input__error")];

  let allErrorsTexts = allErrors.map((err) => err.textContent);

  expect(allErrorsTexts.length).toBe(2);
});

test("Uset fullfill form, submit it and there is no errors to be displayed. Fields keep same value as before form was submitted", async () => {
  const { form, priceInput, timeInput, currencyInput, rateSource } = setup();

  userEvent.type(priceInput, "20");
  userEvent.type(timeInput, "30:30");
  userEvent.selectOptions(currencyInput, ["USD"]);
  userEvent.selectOptions(rateSource, ["Manual"]);

  await waitFor(() => {
    fireEvent.submit(form);
  });

  let allErrors = [...form.querySelectorAll(".input__error")];

  let allErrorsTexts = allErrors.map((err) => err.textContent);

  expect(allErrorsTexts.length).toBe(0);
  expect(priceInput.value).toBe("20");
  expect(timeInput.value).toBe("30:30");
  expect(currencyInput.value).toBe("USD");
  expect(rateSource.value).toBe("Manual");
});
