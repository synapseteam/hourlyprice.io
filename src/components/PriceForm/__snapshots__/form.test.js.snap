// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component snapshot test 1`] = `
<form
  className="form css-8fe1xa-form"
  onSubmit={[Function]}
>
  <div
    className="form__left-column"
  >
    <label
      className="input__label css-18g99wt-label"
    >
      Price
      :
      <input
        autoComplete="off"
        className="input css-15b9r8k-input"
        inputMode="decimal"
        name="price"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="20.30"
        type="text"
      />
    </label>
    <label
      className="input__label css-18g99wt-label"
    >
      Time
      :
      <input
        autoComplete="off"
        className="input css-15b9r8k-input"
        inputMode="decimal"
        name="time"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="HH:MM"
        type="text"
      />
    </label>
    <label
      className="select__label css-18g99wt-label"
    >
      Currency
      :
      <select
        className="select css-r0acw-select"
        name="currency"
        onBlur={[Function]}
        onChange={[Function]}
        value="USD"
      >
        <option
          className="select__option"
          value="USD"
        >
          USD
        </option>
        <option
          className="select__option"
          value="EUR"
        >
          EUR
        </option>
        <option
          className="select__option"
          value="UAH"
        >
          UAH
        </option>
        <option
          className="select__option"
          value="RUB"
        >
          RUB
        </option>
      </select>
    </label>
    <label
      className="select__label css-18g99wt-label"
    >
      Exchange rate
      :
      <select
        className="select css-r0acw-select"
        name="ratesSource"
        onBlur={[Function]}
        onChange={[Function]}
        value="MasterCard"
      >
        <option
          className="select__option"
          value="MasterCard"
        >
          MasterCard
        </option>
        <option
          className="select__option"
          value="Manual"
        >
          Manual
        </option>
      </select>
    </label>
  </div>
  <div
    className="form__right-column"
  />
</form>
`;
